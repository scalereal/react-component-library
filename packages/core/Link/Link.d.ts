import * as React from 'react';
import { WithStyle } from '../../utils';
import { LinkProps } from './types';
export declare const linkSizes: LinkProps['linkSize'][];
export declare const Link: React.FC<LinkProps> & WithStyle;
export default Link;
