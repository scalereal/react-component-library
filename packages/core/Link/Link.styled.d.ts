import { LinkProps } from './types';
export declare const LinkStyled: import("styled-components").StyledComponent<"a", any, LinkProps, never>;
