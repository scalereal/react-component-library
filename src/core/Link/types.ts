import React, { RefObject  } from 'react';
import { AnyStyledComponent } from 'styled-components';

export interface WithStyle {
    Style?: AnyStyledComponent;
}

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type HTMLPropsReact<T> = Omit<React.HTMLProps<T>, 'ref' | 'as'>;

export interface HTMLProps<T> extends HTMLPropsReact<T> {
    ref?: ((instance: T) => void) | RefObject<T>;
}
export interface WithThemeProp {
    // eslint-disable-next-line no-undef
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}
export interface LinkProps extends HTMLProps<HTMLAnchorElement> {
    /** Link url */
    href: string;
}
