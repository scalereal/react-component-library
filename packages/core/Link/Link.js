var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React from 'react';
import Text from '../Text/Text';
import * as Styled from './Link.styled';
var Link = React.memo(React.forwardRef(function (_a, ref) {
    var href = _a.href, linkSize = _a.linkSize, linkColor = _a.linkColor, hoverColor = _a.hoverColor, activeColor = _a.activeColor, props = __rest(_a, ["href", "linkSize", "linkColor", "hoverColor", "activeColor"]);
    var isValidStringOrNumber = function (e) { return typeof e === 'string' || typeof e === 'number'; };
    return (React.createElement(Styled.LinkStyled, __assign({ href: href, ref: ref, linkSize: linkSize, linkColor: linkColor, hoverColor: hoverColor, activeColor: activeColor }, props), React.Children.map(props.children, function (c) {
        return isValidStringOrNumber(c) ? React.createElement(Text, { textSize: linkSize, textColor: linkColor, hoverColor: hoverColor, activeColor: activeColor }, c) : c;
    })));
}));
Link.displayName = 'Link';
Link.Style = Styled.LinkStyled;
Link.defaultProps = {
    linkSize: 'S'
};
export default Link;
