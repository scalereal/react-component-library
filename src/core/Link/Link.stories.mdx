import Link, { linkSizes } from './Link';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Link"
    component={Link}
    argTypes={{
        linkColor: {
            control: 'color'
        },
        hoverColor: {
            control: 'color'
        },
        activeColor: {
            control: 'color'
        },
        linkSize: {
            control: {
                type: 'select',
                options: Object.values(linkSizes)
            }
        }
    }}
/>

# Link

Displays a Link

export const Template = ({ linkSize, linkColor, hoverColor, activeColor, ...args }) => {
    return (
        <Link href="#" linkSize={linkSize} linkColor={linkColor} hoverColor={hoverColor} activeColor={activeColor} {...args}>
            View Link
        </Link>
    );
};

<Preview>
    <Story
        name="Link"
        args={{
            linkSize: 'S'
        }}
    >
        {Template.bind({})}
    </Story>
</Preview>

### Color Variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Link href="#" linkColor="blue" hoverColor="red">
            View Link
        </Link>
        <Link href="#" linkColor="red" hoverColor="blue">
            View Link
        </Link>
    </div>
</Preview>

### Sizes

There are five sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Link href="#" linkSize="XS" linkColor="blue">
            View Link
        </Link>
        <Link href="#" linkSize="S" linkColor="blue">
            View Link
        </Link>
        <Link href="#" linkSize="M" linkColor="blue">
            View Link
        </Link>
        <Link href="#" linkSize="L" linkColor="blue">
            View Link
        </Link>
        <Link href="#" linkSize="XL" linkColor="blue">
            View Link
        </Link>
    </div>
</Preview>

### Props

<Props story="Link" />
