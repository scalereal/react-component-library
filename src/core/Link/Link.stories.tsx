import { Meta, Story } from '@storybook/react/types-6-0'
import React from 'react'
import Link  from './Link'
import { LinkProps } from './types'

const linkSizes : LinkProps['linkSize'][] = ['XS', 'S', 'M', 'L', 'XL']

export default {
    title: 'Components/Core/Link',
    component: Link,
    argTypes: {
        linkColor: {
            control: 'color'
        },
        hoverColor: {
            control: 'color'
        },
        activeColor: {
            control: 'color'
        },
        linkSize:{
            control: {
                type: 'select',
                options: Object.values(linkSizes),
            },
        }
    }
} as Meta;

const Template:Story = ({linkSize,linkColor,hoverColor,activeColor,...args}) => {
    return (
    <Link href="#" linkSize={linkSize} linkColor={linkColor} hoverColor={hoverColor} activeColor={activeColor} {...args}>View Link</Link>
    )
}

export const Default = Template.bind({});
Default.args = {
    linkSize:'S'
}

