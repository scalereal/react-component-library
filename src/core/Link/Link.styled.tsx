import styled from "styled-components"
import linkTheme from "../../theme/core/link/index"
import { LinkProps } from './types';

export const LinkStyled = styled('a')<LinkProps>`
    color: ${({ theme, linkColor }) => (linkColor ? linkColor : theme.defaults.linkColor)};
    background-color: transparent;
    text-decoration: none;
    font-size: ${({ theme, linkSize }) => (theme.sizes[linkSize].fontSize)};
    &:active {
        color: ${({ theme, activeColor }) => (activeColor ? activeColor : theme.defaults.activeColor)};
    }

    &:hover {
        color: ${({ theme, hoverColor }) => (hoverColor ? hoverColor : theme.defaults.hoverColor)};
        text-decoration: underline;
        outline-width: 0;
        cursor: pointer;
    }
`;

LinkStyled.defaultProps = {
    theme:linkTheme
}