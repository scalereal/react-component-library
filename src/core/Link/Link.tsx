import React, { FC, PropsWithChildren } from 'react';
import Text from '../Text/Text';
import * as Styled from './Link.styled';
import { LinkProps } from './types';

export const Link: FC<LinkProps> = ({ href, children, ...props }:PropsWithChildren<LinkProps>) =>{
    const isValidStringOrNumber = (e: any) => typeof e === 'string' || typeof e === 'number';
        return (
            <Styled.LinkStyled href={href} {...props}>
                {React.Children.map(children, c => {
                    return isValidStringOrNumber(c) ? <Text>{c}</Text> : c;
                })}
            </Styled.LinkStyled>
        );
}
export default Link