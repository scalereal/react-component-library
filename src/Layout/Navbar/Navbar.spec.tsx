import { render } from '@testing-library/react';
import * as React from 'react';
import { Navbar } from '.';
import { Link } from '../../core/Link/Link';

describe('NavBar component', () => {
    beforeAll(() => {
        Object.defineProperty(window, "matchMedia", {
          writable: true,
          value: jest.fn().mockImplementation(query => ({
            matches: false,
            media: query,
            onchange: null,
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
          }))
        });
    });

    it('should render with default theme', () => {
        const { container } = render(<Navbar/>);
        expect(container).toMatchSnapshot();
    });
    it('should render with all the props given', () => {
        const { container } = render(
            <Navbar
            navOptions={[
                {
                    option: (
                        <Link href="#" linkSize="S">
                            Home
                        </Link>
                    )
                },
                {
                    option: (
                        <Link href="#" linkSize="S">
                            About Us
                        </Link>
                    )
                },
                {
                    option: (
                        <Link href="#" linkSize="S">
                            Articles
                        </Link>
                    )
                },
                {
                    option: (
                        <Link href="#" linkSize="S">
                            Contact
                        </Link>
                    )
                }]}            />
        );
        expect(container).toMatchSnapshot();
    });
});