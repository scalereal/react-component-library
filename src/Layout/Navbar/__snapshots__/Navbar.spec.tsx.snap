// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar component should render with all the props given 1`] = `
<div>
  <div
    class="sc-hKgILt ciEbKA"
  >
    <div
      class="sc-gsTCUz jTFECS"
    >
      <div>
        <a
          class="sc-jSgupP eMyKUQ"
          href="#"
        >
          <span
            class="sc-eCssSg hBpHdJ"
          >
            Home
          </span>
        </a>
      </div>
      <div>
        <a
          class="sc-jSgupP eMyKUQ"
          href="#"
        >
          <span
            class="sc-eCssSg hBpHdJ"
          >
            About Us
          </span>
        </a>
      </div>
      <div>
        <a
          class="sc-jSgupP eMyKUQ"
          href="#"
        >
          <span
            class="sc-eCssSg hBpHdJ"
          >
            Articles
          </span>
        </a>
      </div>
      <div>
        <a
          class="sc-jSgupP eMyKUQ"
          href="#"
        >
          <span
            class="sc-eCssSg hBpHdJ"
          >
            Contact
          </span>
        </a>
      </div>
    </div>
    <button
      class="sc-dlfnbm hGVZlN Burger"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-bars fa-w-14 "
        data-icon="bars"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`NavBar component should render with default theme 1`] = `
<div>
  <div
    class="sc-hKgILt ciEbKA"
  >
    <div
      class="sc-gsTCUz fcBDvu"
    />
    <button
      class="sc-dlfnbm hGVZlN Burger"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-bars fa-w-14 "
        data-icon="bars"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
